### PR TITLE
BPS-683 Add s2s validation

### PR DESCRIPTION
**Pull request by aliveni**
### Change description ###
duplicate pr of #25 
pr #25 and duplicated pr #27 were closed due to the build pipeline issue.

### CODE REVIEW

File: src/main/resources/application.yaml [hunk 1]
```diff
@@ -12,3 +12,11 @@ management:
...
   application:
     name: Bulk Scan CCD Event Handler Sample Application
+
+idam:
+  s2s-auth:
+    url: ${S2S_URL:http://localhost:4552}
+    secret: ${S2S_SECRET:AAAAAAAAAAAAAAAA}
+    name: ${S2S_NAME:bulk_scan}
+
+allowed-services: bulk_scan_processor,bulk_scan_processor_test
```
Comment (rkondratowicz) [inline src/main/resources/application.yaml]: again, list change missing

Comment (aliveni) [reply src/main/resources/application.yaml]: It's not missing this time 🙂 
Yaml list property is not working with `@Value` annotation. To keep it simple, I added it as comma-separated values.

File: src/main/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/config/AuthConfig.java [hunk 2]
```diff
@@ -9,7 +9,6 @@
 
 @Configuration
 public class AuthConfig {
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/config/AuthConfig.java]: I think below bean config is missing ?

```
@Bean
    @ConditionalOnProperty(name = "idam.s2s-auth.url")
    public AuthTokenValidator tokenValidator(ServiceAuthorisationApi s2sApi) {
        return new ServiceAuthTokenValidator(s2sApi);
    }
```

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/config/AuthConfig.java]: If you rebase and accept the master version of this file, it should be fine.

Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/config/AuthConfig.java]: @lgonczar I still have to add the bean config as @nitinprabhu mentioned above. We need this bean to initiallise `AuthTokenValidator` in `AuthService.java`

File: src/main/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/config/AuthConfig.java [hunk 3]
```diff
@@ -18,4 +20,10 @@ public AuthTokenGenerator authTokenGenerator(
     ) {
         return AuthTokenGeneratorFactory.createDefaultGenerator(secret, name, serviceAuthorisationApi);
     }
+
+    @Bean
+    @ConditionalOnProperty(name = "idam.s2s-auth.url")
```
[the remark this answers is not in the document]
Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/config/AuthConfig.java]: You are correct more bean definitions needs to be added :)

File: src/test/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/controllers/OcrValidationControllerTest.java [hunk 4]
```diff
@@ -32,10 +36,60 @@
     @MockBean
     private OcrDataValidator ocrDataValidator;
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/controllers/OcrValidationControllerTest.java]: Not related with this PR, but we should also verify that the controller returns 200 when request processing results with validation errors (for well-formed requests).

Comment (lgonczar) [reply src/test/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/controllers/OcrValidationControllerTest.java]: Also, not sure if we should return 400 for empty OCR field list - I think this should be left for the validator.

Comment (aliveni) [reply src/test/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/controllers/OcrValidationControllerTest.java]: I'll raise a new pr for the above

Comment (aliveni) [reply src/test/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/controllers/OcrValidationControllerTest.java]: checked with @nitinprabhu about the empty OCR field list, it shouldn't be 400 response.

File: src/main/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/config/AuthConfig.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/config/AuthConfig.java]: added bean definitions

File: src/main/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/services/AuthService.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.authorisation.validators.AuthTokenValidator;
+import uk.gov.hmcts.reform.bulkscanccdeventhandler.services.exception.ForbiddenException;
+import uk.gov.hmcts.reform.bulkscanccdeventhandler.services.exception.UnauthenticatedException;
+
+import static java.util.Arrays.asList;
+import static org.apache.commons.lang.StringUtils.isBlank;
+
+@Component
```
[the remark this answers is not in the document]
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/bulkscanccdeventhandler/services/AuthService.java]: yeah, changed it

File: src/main/resources/application.yaml (hunk 1 above)
Comment (aliveni) [reply src/main/resources/application.yaml]: done

**Comment by aliveni** [comment]
> Good to me. Would be great to have service/controller PR split. For finer grain quality of code and review ;). I can imagine at least some integration testing on controller behalf is better than unit test

Thank you @doncem! I have integration tests ready to be added in the new pr.

**Pull request merged**